client/state/themes/actions.js

```diff
@@ -60,10 +60,35 @@ export function fetchNextPage( site ) {
 	};
 }
 
-export function query( params ) {
-	return {
-		type: THEMES_QUERY,
-		params: params
+function haveQueryParamsChanged( oldQueryParams, newQueryParams ) {
+	const relevantParams = [ 'search', 'filter', 'tier', 'perPage' ];
+	return relevantParams.filter(
+		param => oldQueryParams[ param ] !== newQueryParams[ param ]
+	).length > 0;
+}
+
+/**
+ * If the passed-in query params differ from the current ones, reset the query
+ * state using the new params and fetch the first page of data
+ *
+ * @return {Object}  Promise representing the first page of data fetched, or an empty object if no fetch is made
+ */
+export function query( { params, site } ) {
```

Comment:
This is the biggest change in the PR, revamping the `query` action creator's contract. Before, it always dispatched a `THEMES_QUERY` action with your new query parameters. The biggest problem with this is, the reducer responds to `THEMES_QUERY` by resetting the page count and current themes list.

This solution assumes that if the query params haven't changed, the page and list data is valid and also shouldn't change. So it does a noop if the params we care about like `search` and `filter` are all the same. If the params _have_ changed, it fires `THEMES_QUERY` which resets the query data and empties the data list, and finishes by fetching the first page of data.

To be honest, I'm not sure if this is the best place to implement this logic. But I like it here because it protects all consumers of `query` from accidentally blowing away valid data.